test/matrix/matrix.cpp

```diff
@@ -0,0 +1,782 @@
+    // The strategies have issues with zero rows
+    /*
+    #if defined(GKO_COMPILING_CUDA) || defined(GKO_COMPILING_HIP) || \
+        defined(GKO_COMPILING_DPCPP)
+        CsrWithClassicalStrategy, CsrWithMergePathStrategy,
+        CsrWithSparselibStrategy, CsrWithLoadBalanceStrategy,
+        CsrWithAutomaticalStrategy,
+    #endif
+    */
+    Ell,
+    // Fbcsr is slightly broken
+    /*FbcsrBlocksize1, FbcsrBlocksize2,*/
+    SellpDefaultParameters, Sellp32Factor2, HybridDefaultStrategy,
+    HybridColumnLimitStrategy, HybridImbalanceLimitStrategy,
+    HybridImbalanceBoundedLimitStrategy, HybridMinStorageStrategy,
+    HybridAutomaticStrategy>;
+
+TYPED_TEST_SUITE(Matrix, MatrixTypes, TypenameNameGenerator);
+
+
+TYPED_TEST(Matrix, SpMVIsEquivalentToRef)
+{
+    this->forall_matrix_scenarios([&](auto mtx) {
+        this->forall_vector_scenarios(mtx, [&](auto b, auto x) {
+            mtx.ref->apply(b.ref.get(), x.ref.get());
+            mtx.dev->apply(b.dev.get(), x.dev.get());
+
+            GKO_ASSERT_MTX_NEAR(x.ref, x.dev, this->tol());
+        });
+    });
+}
+
+
+TYPED_TEST(Matrix, AdvancedSpMVIsEquivalentToRef)
+{
+    this->forall_matrix_scenarios([&](auto mtx) {
+        this->forall_vector_scenarios(mtx, [&](auto b, auto x) {
+            auto alpha = this->gen_scalar();
+            auto beta = this->gen_scalar();
+
+            mtx.ref->apply(alpha.ref.get(), b.ref.get(), alpha.ref.get(),
+                           x.ref.get());
+            mtx.dev->apply(alpha.dev.get(), b.dev.get(), alpha.dev.get(),
+                           x.dev.get());
+
+            GKO_ASSERT_MTX_NEAR(x.ref, x.dev, this->tol());
+        });
+    });
+}
+
+
+TYPED_TEST(Matrix, MixedSpMVIsEquivalentToRef)
+{
+    using MixedVec = typename TestFixture::MixedVec;
+    this->forall_matrix_scenarios([&](auto mtx) {
+        this->template forall_vector_scenarios<MixedVec>(
+            mtx, [&](auto b, auto x) {
+                mtx.ref->apply(b.ref.get(), x.ref.get());
+                mtx.dev->apply(b.dev.get(), x.dev.get());
+
+                GKO_ASSERT_MTX_NEAR(x.ref, x.dev, this->mixed_tol());
+            });
+    });
+}
+
+
+TYPED_TEST(Matrix, MixedAdvancedSpMVIsEquivalentToRef)
+{
+    using MixedVec = typename TestFixture::MixedVec;
+    this->forall_matrix_scenarios([&](auto mtx) {
+        this->template forall_vector_scenarios<MixedVec>(
+            mtx, [&](auto b, auto x) {
+                auto alpha = this->template gen_scalar<MixedVec>();
+                auto beta = this->template gen_scalar<MixedVec>();
+
+                mtx.ref->apply(alpha.ref.get(), b.ref.get(), alpha.ref.get(),
+                               x.ref.get());
+                mtx.dev->apply(alpha.dev.get(), b.dev.get(), alpha.dev.get(),
+                               x.dev.get());
+
+                GKO_ASSERT_MTX_NEAR(x.ref, x.dev, this->mixed_tol());
+            });
+    });
+}
+
+
+TYPED_TEST(Matrix, ConvertToCsrIsEquivalentToRef)
```

Comment:
Should we be testing conversions for other matrix formats as well ?

Reply:
Which formats do you mean? We only have conversions from all formats to Csr/Dense and from Csr/Dense to all formats, with minor exceptions

Reply:
Ah, yes. That is true. I forgot that you only test if all the permutations for a certain format have been implemented.

Reply:
Do we also make the minor exception test shown here?

Reply:
The tests are already handled inside the format-specific test files, so I will leave them out for now, but we can definitely add such tests later.